Update runtime to 5.15-21.08

tinywrkb opened the pull request:
Marked as draft as this should be merged into a new `5.15-21.08` branch.  
Try and see if the base app can be built against the `5.15-21.08` runtime without further changes, and create a test build to try, as build time is too long to do this locally.

Add the following patches:
 - glibc 2.33: fixes rendering with some sites, see https://bugzilla.redhat.com/show_bug.cgi?id=1904652.
 - harfbuzz 3.0.0: fixes compilation
 - icu 68: fix compilation of in-tree libxml2 as TRUE/FALSE macros were removed, see https://icu.unicode.org/download/68.
   runtime libxml2 cannot be used as icu support is disabled, reported at https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/issues/1336.
 - disable thin archives: fixes linking of libbrowser.a. apparently a binutils regresssion (thanks @refi64), see https://sourceware.org/bugzilla/show_bug.cgi?id=28138.

 The glib & harfbuzz patches were imported from the Arch Linux packaging of qt5-webengine.

tinywrkb committed:
Update runtime to 5.15-21.08

Add the following patches:
 - glibc 2.33: fixes rendering with some sites, see https://bugzilla.redhat.com/show_bug.cgi?id=1904652.
 - harfbuzz 3.0.0: fixes compilation
 - icu 68: fix compilation of in-tree libxml2 as TRUE/FALSE macros were removed, see https://icu.unicode.org/download/68.
   runtime libxml2 cannot be used as icu support is disabled, reported at https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/issues/1336.
 - disable thin archives: fixes linking of libbrowser.a. apparently a binutils regresssion (thanks @refi64), see https://sourceware.org/bugzilla/show_bug.cgi?id=28138.

 The glib & harfbuzz patches were imported from the Arch Linux packaging of qt5-webengine.

tinywrkb commented:
Note that we need to request that the default branch should be set as `5.15-21.08`, so flathubbot will run f-e-d-c.

tinywrkb commented on io.qt.qtwebengine.BaseApp.json:
```diff
@@ -37,6 +37,18 @@
+                {
+                    "name": "icu67",
+                    "subdir": "source",
+                    "no-autogen": true,
```
`"no-autogen": true,` is actually unneeded, but I'll wait for the current CI to end before pushing a change.

Erick555 commented:
In Archlinux qt5-webengine seems to build fine with latest icu. Do you know why here it's different?

tinywrkb commented:
@Erick555 ~~I believe the Arch Linux package builds from the latest commit in 87-based branch of the qtwebengine-chromium git submodule.~~  
~~We just default to whatever commit the submodule was set to when the qtwebengine release was tagged.
Maybe it's a good idea to follow Arch considering all those backports in the 87-branch.~~

~~I'll try and look at this again tomorrow.~~

Oh, I was wrong.

tinywrkb committed:
f-e-d-c: set the upstream source

tinywrkb commented:
In this Flatpak

```
Running configuration tests...
Checking for compatible jpeglib... yes
Checking for libevent... no
Checking for compatible libxml2 and libxslt... no
Checking for minizip... no
Checking for re2... no
Checking for snappy... no


Qt WebEngine Build Tools:
  Optional system libraries used:
    re2 .................................. no
    snappy ............................... no
    minizip .............................. no
    libevent ............................. no
    libxml2 and libxslt .................. no

Qt WebEngineCore:
  Kerberos Authentication ................ no
```

In Arch Linux

```
Running configuration tests...
Checking for compatible jpeglib... jyes
Checking for libevent... yes
Checking for compatible libxml2 and libxslt... yes
Checking for minizip... yes
Checking for re2... yes
Checking for snappy... yes

Qt WebEngine Build Tools:
  Optional system libraries used:
    re2 .................................. yes
    snappy ............................... yes
    minizip .............................. yes
    libevent ............................. yes
    libxml2 and libxslt .................. yes
Qt WebEngineCore:
  Kerberos Authentication ................ yes
```

Erick555 commented:
Hm but Arch ships exactly same version of libxml2 and libxslt as 21.08 sdk:

https://github.com/archlinux/svntogit-packages/blob/packages/libxml2/trunk/PKGBUILD#L7

https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/blob/release/21.08/elements/components/libxml2.bst#L34

https://github.com/archlinux/svntogit-packages/blob/packages/libxslt/trunk/PKGBUILD#L5

https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/blob/release/21.08/elements/components/libxslt.bst#L36
Maybe build script doesn't look in right place? How it was in 20.08?

tinywrkb commented:
@Erick555 

20.08 ships ICU 67, that's [before the TRUE and FALSE macros removal](https://icu.unicode.org/download/68). **edit: this change broke the embedded libxml2 compilation.**

```
$ grep U_ICU_VERSION_MAJOR_NUM /var/lib/flatpak/runtime/org.freedesktop.Sdk/x86_64/20.08/active/files/include/unicode/uvernum.h

#define U_ICU_VERSION_MAJOR_NUM 67
```

21.08 ships ICU 69

```
$ grep U_ICU_VERSION_MAJOR_NUM /var/lib/flatpak/runtime/org.freedesktop.Sdk/x86_64/21.08/active/files/include/unicode/uvernum.h

#define U_ICU_VERSION_MAJOR_NUM 69
```

If I understand it correctly, the [libxml2 configure test checks if libxml2 was built with ICU support enabled](https://invent.kde.org/qt/qt/qtwebengine/-/blob/v5.15.6-lts/src/buildtools/configure.json#L199), and it's not enabled in the SDK.

`/var/lib/flatpak/runtime/org.freedesktop.Sdk/x86_64/{20.08,21.08}/active/files/include/libxml2/libxml/xmlversion.h`
```
#if 0
#define LIBXML_ICU_ENABLED
#endif
```

tinywrkb commented:
~~I'm not sure why, but when building locally and setting the symbol level to 0 gave me a more useful error than the linker failure.~~   
~~I haven't confirmed yet this actually help to complete the build, my system is still compiling.~~  
Scratch that, I noticed the runtime recently updated harfbuzz.

tinywrkb commented:
Disabling jumbo builds moved the linker error from `src/core/release/obj/content/browser/libbrowser.a` to `src/core/release/obj/third_party/blink/renderer/core/editing/libediting.a`.

tinywrkb commented:
~~Is it possible that there's an issue with GCC?~~ **edit: NVM**

```
$ file $(ar -t libbrowser.a)

browser/about_url_loader_factory.o:           ELF 64-bit LSB relocatable, x86-64, version 1 (SYSV), not stripped
browser/accessibility_event_recorder.o:       ELF 64-bit LSB relocatable, x86-64, version 1 (SYSV), not stripped
browser/accessibility_tree_formatter_base.o:  ELF 64-bit LSB relocatable, x86-64, version 1 (SYSV), not stripped
browser/accessibility_tree_formatter_blink.o: ELF 64-bit LSB relocatable, x86-64, version 1 (GNU/Linux), not stripped
browser/accessibility_tree_formatter_stub.o:  ELF 64-bit LSB relocatable, x86-64, version 1 (SYSV), not stripped
browser/browser_accessibility.o:              ELF 64-bit LSB relocatable, x86-64, version 1 (SYSV), not stripped
browser/browser_accessibility_manager.o:      ELF 64-bit LSB relocatable, x86-64, version 1 (GNU/Linux), not stripped
browser/browser_accessibility_position.o:     ELF 64-bit LSB relocatable, x86-64, version 1 (SYSV), not stripped
browser/browser_accessibility_state_impl.o:   ELF 64-bit LSB relocatable, x86-64, version 1 (GNU/Linux), not stripped
browser/one_shot_accessibility_tree_search.o: ELF 64-bit LSB relocatable, x86-64, version 1 (SYSV), not stripped
browser/after_startup_task_utils.o:           ELF 64-bit LSB relocatable, x86-64, version 1 (SYSV), not stripped
browser/appcache.o:                           ELF 64-bit LSB relocatable, x86-64, version 1 (GNU/Linux), not stripped
...
```

tinywrkb commented:
Probably not the correct solution, but it finally builds, or at least locally.

tinywrkb commented:
qutebrowser is working with the 5.15-21.08 base app, but it looks like we might need the glibc 2.33 patch. **edit: I can confirm that the rendering issue is solved when the patch is applied.**

tinywrkb commented:
[Bug report for enabling ICU support in libxml2](https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/issues/1336).

tinywrkb commented:
I squashed the commits, and changed the sed calls to patches.  
If this still builds correctly, then maybe we can already pull this into a new `5.15-21.08` branch, but it would be nice if we could:
1. Find a cleaner way to disable the thin archive option.
2. Discover the reason why thin archive breaks linking, and fix the issue.

@aleixpol can you please review this?
@refi64 do you have some insight on why linking is failing when thin archive is enabled?

refi64 commented:
@tinywrkb After looking into this a bit, I believe it's [this bug](https://sourceware.org/bugzilla/show_bug.cgi?id=28138) which *sounds* like the core issue. It's been fixed in the sources, but I'm not sure if it's made it into a release yet.

In the mean time, the smallest possible patch to disable thin archives would be to edit [these lines](https://source.chromium.org/chromium/chromium/src/+/main:build/config/compiler/BUILD.gn;l=1750-1760?q=thin_archive%20language:gn&ss=chromium%2Fchromium%2Fsrc&start=11) and change `arflags = [ "-T" ]` to `arflags = []` (or just remove the whole body of the `config(...)` block altogether).

tinywrkb commented:
@refi64 Thanks! I really appreciate it.

tinywrkb commented:
@aleixpol I think this is ready to be pulled-in into a new 5.15-21.08 branch.

I'll try later, in a new PR, building Chromium from the latest commit in the `87-based` branch, which might have some vulnerabilities patched.

aleixpol commented:
Can you remove the WIP tag then?

tinywrkb commented:
> Can you remove the WIP tag then?

Done!

aleixpol commented:
This should be targetting 5.15-21.08.
We probably don't want to switch runtimes to every user of the runtime.

tinywrkb commented:
> This should be targetting 5.15-21.08.

First line in the first post.
> Marked as draft as this should be merged into a new 5.15-21.08 branch.

First line in my last comment requesting to be reviewed.

> @aleixpol I think this is ready to be pulled-in into a **new 5.15-21.08 branch.**

tinywrkb commented:
bot, build io.qt.qtwebengine.BaseApp

aleixpol commented:
Thanks!
